test/libraries/ProtocolFeeLibrary.t.sol

```diff
@@ -39,35 +36,47 @@ contract ProtocolFeeLibraryTest is Test {
-        lpFee = uint24(bound(lpFee, 0, LPFeeLibrary.MAX_LP_FEE));
-        self = uint24(bound(self, 0, ProtocolFeeLibrary.MAX_PROTOCOL_FEE));
-        assertGe(ProtocolFeeLibrary.calculateSwapFee(self, lpFee), lpFee);
+        fee = uint24(ProtocolFeeLibrary.MAX_PROTOCOL_FEE - 1) << 12 | ProtocolFeeLibrary.MAX_PROTOCOL_FEE - 1;
+        assertTrue(ProtocolFeeLibrary.isValidProtocolFee(fee));
+
+        fee = uint24(0) << 12 | uint24(0);
+        assertTrue(ProtocolFeeLibrary.isValidProtocolFee(fee));
     }
 
-    function test_fuzz_calculateSwapFeeNeverEqualsMax(uint24 self, uint24 lpFee) public pure {
-        lpFee = uint24(bound(lpFee, 0, LPFeeLibrary.MAX_LP_FEE - 1));
-        self = uint24(bound(self, 0, ProtocolFeeLibrary.MAX_PROTOCOL_FEE));
-        assertLt(ProtocolFeeLibrary.calculateSwapFee(self, lpFee), LPFeeLibrary.MAX_LP_FEE);
+    function test_fuzz_isValidProtocolFee(uint24 fee) public pure {
+        if ((fee >> 12 > ProtocolFeeLibrary.MAX_PROTOCOL_FEE) || (fee % 4096 > ProtocolFeeLibrary.MAX_PROTOCOL_FEE)) {
+            assertFalse(ProtocolFeeLibrary.isValidProtocolFee(fee));
+        } else {
+            assertTrue(ProtocolFeeLibrary.isValidProtocolFee(fee));
+        }
     }
 
     function test_calculateSwapFee() public pure {
-        uint24 self = uint24(ProtocolFeeLibrary.MAX_PROTOCOL_FEE);
-        uint24 lpFee = LPFeeLibrary.MAX_LP_FEE;
-        assertEq(ProtocolFeeLibrary.calculateSwapFee(self, lpFee), LPFeeLibrary.MAX_LP_FEE);
-
-        lpFee = 3000;
-        assertEq(ProtocolFeeLibrary.calculateSwapFee(self, lpFee), 3997);
-
-        lpFee = 0;
-        assertEq(ProtocolFeeLibrary.calculateSwapFee(self, lpFee), ProtocolFeeLibrary.MAX_PROTOCOL_FEE);
+        assertEq(
+            ProtocolFeeLibrary.calculateSwapFee(uint24(ProtocolFeeLibrary.MAX_PROTOCOL_FEE), LPFeeLibrary.MAX_LP_FEE),
+            LPFeeLibrary.MAX_LP_FEE
+        );
+        assertEq(ProtocolFeeLibrary.calculateSwapFee(uint24(ProtocolFeeLibrary.MAX_PROTOCOL_FEE), 3000), 3997);
+        assertEq(
+            ProtocolFeeLibrary.calculateSwapFee(uint24(ProtocolFeeLibrary.MAX_PROTOCOL_FEE), 0),
+            ProtocolFeeLibrary.MAX_PROTOCOL_FEE
+        );
+        assertEq(ProtocolFeeLibrary.calculateSwapFee(0, 0), 0);
+        assertEq(ProtocolFeeLibrary.calculateSwapFee(0, 1000), 1000);
+    }
 
-        self = 0;
-        assertEq(ProtocolFeeLibrary.calculateSwapFee(self, lpFee), 0);
+    function test_fuzz_calculateSwapFee(uint24 protocolFee, uint24 lpFee) public pure {
+        protocolFee = uint24(bound(protocolFee, 0, ProtocolFeeLibrary.MAX_PROTOCOL_FEE));
+        lpFee = uint24(bound(lpFee, 0, LPFeeLibrary.MAX_LP_FEE));
+        // if lp fee is not the max, the swap fee should never be the max since the protocol fee is taken off first and then the lp fee is taken from the remaining amount
+        if (lpFee < LPFeeLibrary.MAX_LP_FEE) {
+            assertLt(ProtocolFeeLibrary.calculateSwapFee(protocolFee, lpFee), LPFeeLibrary.MAX_LP_FEE);
```

Comment:
1 final thing you should just calculate `ProtocolFeeLibrary.calculateSwapFee(protocolFee, lpFee)` one time currently its called 3 times.

Reply:
ty!